### ISSUE TITLE
[bug] export-pkg yields package ID of Package_ID_unknown in package_revision_mode

### ISSUE THREAD

**Issue by mgaunard** [conan-io/conan#11951]
I use a workflow where I put conanfile.py inside the source of my packages, and typically compile with the following workflow.
 - `mkdir -p build/release && cd build/release`
 - `conan install ../.. -s build_type=Release`
 - `conan build ../..`
 - `conan package ../..`
 - `conan export-pkg --package-folder=package`
 - `conan upload myproject/* --all -c` (this step is actually only done by the CI/CD user, cannot be run locally)
 
I've been trying to move to either of `recipe_revision_mode` or `package_revision_mode`, but only for the packages that I export; not for the dependencies that I import (which, since they're from conancenter, must remain `semver_direct_mode`).

I therefore did the following:
 - `mkdir -p build/release && cd build/release`
 - `CONAN_REVISIONS_ENABLED=1 conan install ../.. -s build_type=Release`
 - `CONAN_REVISIONS_ENABLED=1 CONAN_DEFAULT_PACKAGE_ID_MODE=package_revision_mode conan build ../..`
 - `CONAN_REVISIONS_ENABLED=1 CONAN_DEFAULT_PACKAGE_ID_MODE=package_revision_mode conan package ../..`
 - `CONAN_REVISIONS_ENABLED=1 CONAN_DEFAULT_PACKAGE_ID_MODE=package_revision_mode conan export-pkg --package-folder=package`

Result of the `export-pkg`:
```
Exporting package recipe
myproject/0.1: The stored package has not changed
myproject/0.1: Using git commit as the recipe revision: ac6c0f6901eb3174544e6ce216563626135738d2
myproject/0.1: Exported revision: ac6c0f6901eb3174544e6ce216563626135738d2
Packaging to Package_ID_unknown
myproject/0.1: Exporting to cache existing package from user folder
myproject/0.1: Package folder /home/mgaunard/.conan/data/myproject/0.1/package/Package_ID_unknown
myproject/0.1: Packaged 13 '.h' files
myproject/0.1: Packaged 1 '.txt' file: LICENSE.txt
myproject/0.1: Package 'Package_ID_unknown' created
myproject/0.1: Created package revision bf9d380f49dccfb2140d9db2c7cad8a9
```

It deduces the package ID to be `Package_ID_unknown`, which suggests there is a problem there on how it's updating the graph lock. I tried to look at the code but couldn't quite understand how that works.

Of note, using `recipe_revision_mode` yields what looks like a valid hash, but somehow the package ID doesn't change when the RREV changes, so this doesn't appear to work correctly either.

**Comment by memsharded** [comment]
Hi @mgaunard 

> I've been trying to move to either of recipe_revision_mode or package_revision_mode, but only for the packages that I export; not for the dependencies that I import (which, since they're from conancenter, must remain semver_direct_mode).

This is not possible. The default mode applies to all packages, including those that come from ConanCenter. The result is that packages from ConanCenter will not have pre-compiled binaries available for that mode, and consumers of those packages will not be able to compute their ``package_id`` without first building new binaries for the dependencies, this is why it results in ``Package_ID_unknown``. What you are doing is trying to bypass this with a local build + export_pkg, but that would be undefined behavior.

In any case, ``package_revision_mode`` has been problematic, and it was decided to remove it, and it has already been removed in Conan 2.0, you can check https://github.com/conan-io/tribe/pull/30.
It is completely discouraged to use it in 1.X too. If you want to be strict, the recommended way is ``recipe_revision_mode`` (full mode in 2.0), and package binaries shouldn't be rebuilt if they already have a binary, so the number of ``package_revisions`` should always be 1.

When the recipe revisions of the dependencies change, then you should definitely get a new ``package_id`` for your package. Please check it, and report a new issue for that if necessary. I am marking this as a question, as solved in 2.0, thanks!

**Comment by mgaunard** [comment]
Understood about `package_revision_mode`, you chose to remove it so it doesn't matter. You should reflect the documentation.

However, that doesn't change the bug report at all. As I said, the same sequence of operations with `CONAN_DEFAULT_PACKAGE_ID_MODE=recipe_revision_mode` instead of `CONAN_DEFAULT_PACKAGE_ID_MODE=package_revision_mode` doesn't work, the package ID doesn't change and always stays the same.

Here is for example two outputs with two recipe revisions:
```
Exporting package recipe
myproject/0.1: The stored package has not changed
myproject/0.1: Using git commit as the recipe revision: ac6c0f6901eb3174544e6ce216563626135738d2
myproject/0.1: Exported revision: ac6c0f6901eb3174544e6ce216563626135738d2
myproject/0.1: Forced build from source
Packaging to 452cd7b1ec26b2f71d9dd5b880830c49828cbd6f
myproject/0.1: Exporting to cache existing package from user folder
myproject/0.1: Package folder /home/mgaunard/.conan/data/myproject/0.1/package/452cd7b1ec26b2f71d9dd5b880830c49828cbd6f
myproject/0.1: Packaged 13 '.h' files
myproject/0.1: Packaged 1 '.txt' file: LICENSE.txt
myproject/0.1: Package '452cd7b1ec26b2f71d9dd5b880830c49828cbd6f' created
myproject/0.1: Created package revision fc23f8b65d8c1ec77a582f0283b619e6
```

```
Exporting package recipe
myproject/0.1: The stored package has not changed
myproject/0.1: Using git commit as the recipe revision: c1edc5cbb36480d3b34e6d6c1fcbcf3c853de2ea
myproject/0.1: Exported revision: c1edc5cbb36480d3b34e6d6c1fcbcf3c853de2ea
myproject/0.1: Forced build from source
Packaging to 452cd7b1ec26b2f71d9dd5b880830c49828cbd6f
myproject/0.1: Exporting to cache existing package from user folder
myproject/0.1: Package folder /home/mgaunard/.conan/data/myproject/0.1/package/452cd7b1ec26b2f71d9dd5b880830c49828cbd6f
myproject/0.1: Packaged 13 '.h' files
myproject/0.1: Packaged 1 '.txt' file: LICENSE.txt
myproject/0.1: Package '452cd7b1ec26b2f71d9dd5b880830c49828cbd6f' created
myproject/0.1: Created package revision fc23f8b65d8c1ec77a582f0283b619e6
```

Notice how exported revision is different, yet package ID and package revision remain the same.

The documentation clearly says that you can specify for each dependency which package revision mode to use to compute the package ID. I'd like to use `semver_direct_mode` for external dependencies and `recipe_revision_mode` for my own packages.
The documentation however doesn't say how to choose the package revision mode for the package that is being built/exported. The only way I've found to do that is to use `CONAN_DEFAULT_PACKAGE_ID_MODE`.

At the time of `export-pkg`, the graph lock is already constructed and resolved, but the package being built hasn't been assigned a package ID yet. There is no reason why it should be a problem to generate it with the correct mode there?

Clearly there is something broken in `export-pkg`, hence why this is a bug, not a question.

**Comment by memsharded** [comment]
There are different things here:

- You cannot just apply the ``CONAN_DEFAULT_PACKAGE_ID_MODE=package_revision_mode`` to the ``conan build``, ``conan package`` and ``conan export-pkg``. Whatever default package_id_mode you select has to be configured always, including the ``conan install``, otherwise you will not be able to consume this package later. The recommended way is using configuration in ``conan.conf`` (legacy) or in ``global.conf``.
- If you want to select that a given recipe, for example, ``myproject``, depends in a certain package_id_mode to some or all of its dependencies, the way to do it is with ``package_id()`` method, check: https://docs.conan.io/en/latest/creating_packages/define_abi_compatibility.html#using-package-id-for-package-dependencies
- For ``myproject/0.1`` to get 2 different package_ids, is when its dependencies change. What are the dependencies of ``myproject``? Can you generate a new revision for one of them, and check how the ``package_id`` of ``myproject`` changes?

I have done this unit test, that works correctly:
```python
def test_recipe_revision_mode():
    client = TestClient()
    client.run("config set general.default_package_id_mode=recipe_revision_mode")

    client.save({"libb/conanfile.py": GenConanfile("libb", "0.1"),
                 "app/conanfile.py": GenConanfile("app", "0.1").with_require("libb/0.1")})

    # create a revision for libb
    client.run("create libb")
    assert "libb/0.1: Exported revision: f7d0ff5836a100ded1666a09171046d0" in client.out
    client.run("export-pkg app")
    assert "app/0.1: Package '637aa018b276286c54fe8f939f256493d4dfbc16' created" in client.out

    # Now create a new revision for libb
    client.save({"libb/conanfile.py": str(GenConanfile("libb", "0.1")) + "# comment"})
    client.run("create libb")
    assert "libb/0.1: Exported revision: 560c6a4a16f61b6e0253874abb6b23ee" in client.out
    # The export-pkg will get a new package_id as it depends on a new revision
    client.run("export-pkg app")
    assert "app/0.1: Package 'a010b1231c483d99ec749122e4d692ce4baaf0b5' created" in client.out
```

As you can see, 2 different ``package_ids`` are being created for ``app/0.1`` when its dependency gets a new revision. It is not a bug, it seems it is working correctly.

**Comment by mgaunard** [comment]
Different dependencies need different package ID modes for them. I need conancenter packages to use `semver_direct_mode` and my own packages to use `recipe_revision_mode`. I'm achieving this by keeping the default at `semver_direct_mode` and overriding it for my own `requires` in `package_id()`. Doing it the other way around causes more problems due to `build_requires` for which I cannot specify the mode.

My understanding of `recipe_revision_mode` is that the package ID of a given package does not only depend on its dependencies, but also on its own RREV. Are you saying that isn't the case?
If it's not the case, isn't it entirely useless? I need to make it so that when my source code changes, the package ID changes. I cannot achieve reproducible builds without this.

**Comment by memsharded** [comment]
> Different dependencies need different package ID modes for them. I need conancenter packages to use semver_direct_mode and my own packages to use recipe_revision_mode. I'm achieving this by keeping the default at semver_direct_mode and overriding it for my own requires in package_id(). Doing it the other way around causes more problems due to build_requires for which I cannot specify the mode.

That is fine, but you cannot change that by temporarily changing the global ``default_package_id_mode``. You have to select one global mode and attach to it. Then, the recipes that want to diverge from that default, should use the ``package_id()`` method to define how they depend on other packages. So basically what you want is to define your ``package_id()`` method in your recipes to define the mode for your own packages but not conan-center ones.

> My understanding of recipe_revision_mode is that the package ID of a given package does not only depend on its dependencies, but also on its own RREV. Are you saying that isn't the case?

No, ``recipe_revision_mode`` is only for how the dependencies affect your current package ``package_id``. The current revision of the current package is not factored into the current ``package_id``, but it is another part of the reference, like: ``pkgname/version@user/channel#recipe_revision:package_id``. ``package_id`` are not intended to be globally unique, just unique for a given ``pkgname/version@user/channel#recipe_revision``. The full identifier ``pkgname/version@user/channel#recipe_revision:package_id`` is the unique thing, and it can be perfectly used for reproducible builds (of course, with some limits: https://blog.conan.io/2019/09/02/Deterministic-builds-with-C-C++.html)

> If it's not the case, isn't it entirely useless? I need to make it so that when my source code changes, the package ID changes. I cannot achieve reproducible builds without this.

You will not achieve reproducible builds anyway, if you are using dependencies from ConanCenter as-is, and you are not both locking them and using ``recipe_revision_mode`` for them. Because recipes in ConanCenter change continuously, and you are linking them. If you need fully reproducible builds, you need to depend in production on an immutable copy of ConanCenter packages, and also to use stricter modes for those too.

But yes, ``recipe_revision_mode`` is to achieve that: it guarantees that when the dependencies change, the consumers need to build a new binary

**Comment by dornbirndevelops** [comment]
> In any case, `package_revision_mode` has been problematic, and it was decided to remove it, and it has already been removed in Conan 2.0, you can check [conan-io/tribe#30](https://github.com/conan-io/tribe/pull/30). It is completely discouraged to use it in 1.X too. If you want to be strict, the recommended way is `recipe_revision_mode` (full mode in 2.0), and package binaries shouldn't be rebuilt if they already have a binary, so the number of `package_revisions` should always be 1.

@memsharded would it make sense to have this fact mentioned in the documentation somewhere?

**Comment by memsharded** [comment]
Conan 2.0 is due next month, so closing this as solved in 2.0 (``package_revision_mode`` has been removed, and ``Package_ID_unknown`` does not exist anymore)
